Fixes for diarrhoea generic first appointment logic and updating of individual properties across modules

matt-graham opened the pull request:
Related to #1348  and #1366.

As discussed in software engineering meeting today, this PR addresses two known issues with current refactored generic first appointments logic:

1. A change in the conditions on which the diarrhoea module does or does not perform an action on a generic first appointment introduced accidentally by the refactoring in #1292 is fixed to return to the previous behaviour (not performing any further action if the individual is over the age of 5 _or_ does not have any diarrhoea symptoms).
2. The issue identified by @marghe-molaro in https://github.com/UCL/TLOmodel/issues/1348#issuecomment-2134959827, that is that in 
  https://github.com/UCL/TLOmodel/blob/df25c764e6d6fc99f224a65d18a030c67630c5a4/src/tlo/methods/hsi_generic_first_appts.py#L80-L103
  updates to an individual properties by previous modules in the loop are not reflected necessarily in the `patient_details` structure passed to the `do_at_generic_first_appt` method of the current module in the loop as the updates are only made to the dataframe after the loop has completed. Due the subsequent changes to the `PatientDetails` type made in #1315, this will actually only be an issue if an update is made to a property by a module that has previously been read (and cached) by that or another module and then a subsequent module reads the same property again (as it will get the initial cached value). I'm not sure there are cases where that chain actually happens currently given only very few modules actually output updates to the individual properties and they are for properties usually specific to the module, so this possibly explains why @willGraham01 was able to get consistent dataframe checksums when working on #1366 even without this fix.

The fix for (2) here uses the simple approach I mentioned in https://github.com/UCL/TLOmodel/issues/1348#issuecomment-2139374633 which will loose some of the performance benefits we currently have by doing updates to the dataframe whenever a module returns properties to update (rather than doing all updates at once). In the longer term #1387 should fix the same issue in a more performant manner.

matt-graham commented:
@marghe-molaro this PR should hopefully deal with the issues so once this is reviewed / merged to `master` you should be able to merge in to your branch for analysis runs and re-run to check if we restore original behaviour. If it does then we can close #1348.

marghe-molaro commented:
Thank you very much @matt-graham!
Now that @tamuri has approved these changed I will merge these fixes into commit a91d60c518069ed11477f27fb3c81f1ee92302ca, which produced the plots in which we first spotted an issue. This will allow us to establish whether the issues fixed in this PR were behind the discrepancy.

marghe-molaro commented:
@matt-graham it appears there's an issue with the diarrhoea test

matt-graham commented:
It looks like Will [also updated the tests](https://github.com/UCL/TLOmodel/pull/1366/files#diff-ccf2792fb9d59e7d4b39bc97829dfcf18e72901eb0f2728c88350de927683015) that are failing here in #1366 to force individual to have diarrhoea symptoms - currently investigating why that was needed and will update test here if necessary.

matt-graham commented:
> It looks like Will [also updated the tests](https://github.com/UCL/TLOmodel/pull/1366/files#diff-ccf2792fb9d59e7d4b39bc97829dfcf18e72901eb0f2728c88350de927683015) that are failing here in #1366 to force individual to have diarrhoea symptoms - currently investigating why that was needed and will update test here if necessary.

I realized the manual setting of `diarrhoea` symptoms was now necessary as prior to the refactoring we called a method `Diarrhoea.do_when_presentation_with_diarrhoea` which assumed symptoms had already been checked, while post refactoring the logic of this method has been moved to the  `Diarrhoea.do_at_generic_first_appt` method which now with the fix in the PRs, exits straight away if there isn't `"diarrhoea"` present in the passed set of symptoms. Have now updated the test to manually create a set of symptoms with `"diarrhoea"` present where applicable so this should hopefully fix the failure.

matt-graham commented:
As this is a change to the test to keep it up to sync with the changes in the module, this shouldn't affect the validity of any runs with the changes here merged in before the last commit @marghe-molaro, just in case you've already set anything off!

marghe-molaro commented:
Hi @matt-graham, quick update: @tbhallett noticed that this PR was based on a version of master that had brought in changes to the HIV/TB/malaria modules which lead to a rescaling of DALYs and scheduled HSIs for those diseases (#1273). By comparing the original runs which included the error with the fixes brought in this PR (#1389), it has therefore become difficult to isolate the effect of modules reordering on the changes to the HIV DALYs evolution.

I will submit a new set of test runs from the current version of master where the only change added is the scrambling of the disease modules to recover the order in the original code vs the one introduced by PR #1292. Hopefully this will allow us to establish once and for all whether reordering the modules during the first generic appts leads to any statistically significant changes. Let me know if you have any thoughts on this.
